compiler/rustc_hir_analysis/src/hir_ty_lowering/lint.rs

```diff
@@ -201,14 +220,54 @@ impl<'tcx> dyn HirTyLowerer<'tcx> + '_ {
                 "{pre}use `impl {trait_name}` to return an opaque type, as long as you return a \
                  single underlying type",
             );
-            diag.multipart_suggestion_verbose(msg, impl_sugg, Applicability::MachineApplicable);
+
+            // Six different cases to consider when suggesting `impl Trait` for a return type:
```

Comment:
I special cased `&mut` because the current suggestion for `fn code() -> &mut Trait` looks like:
``` rust
7 | fn code() -> &'static mut Trait {
  |               +++++++

7 | fn code() -> &mut impl Trait {
  |                   ++++
```
(Note: The `'static` lifetime suggestion is from `E0106` and it's complicated to fix that because it happens in `rustc_resolve`.)

I thought the current suggestion might be very confusing for someone new to Rust than this proposed suggestion:
``` rust
error[E0782]: cannot return a mutable reference to a bare trait
  --> $DIR/reference-to-bare-trait-in-fn-inputs-and-outputs-issue-125139.rs:44:25
   |
LL |     fn parrot() -> &mut Trait {
   |                         ^^^^^
   |
help: use `impl Trait` to return an opaque type, as long as you return a single underlying type
   |
LL |     fn parrot() -> impl Trait {
   |                    ~~~~~~~~~~
```

But I agree that special casing for lifetime is more complicated than necessary. Maybe I can keep the `case 6` branch and remove the `cases 4 and 5` branch?

Reply:
I don't think mutability has anything to do with this special case being confusing. Can't I just create the same exact case with an immutable reference?

Reply:
Ah yeah, it would be the same for immutable reference as well.

So can I remove all special casing for return type?